### DISCUSSION TITLE
Handlers reload

### DISCUSSION THREAD

**Discussion by mgvs** [aio-libs/aiohttp#7925] [Q&A]
Is any way to reload handlers without app restart? I have dynamic api part which has to be updated by webhook call without full application restart.

**Comment by webknjaz** [comment]
The routing is frozen upon the app creation, you can't change that. What do you mean by "reload", anyway?

**Reply by mgvs** [reply]
This is what I mean - it is frozen. And no way to add/delete handlers from routing while app working

**Reply by webknjaz** [reply]
That's by design, to avoid weird side effects. Although, I can't imagine why one would need that in the prod mode. Many frameworks offer autoreload of the dev server, but only in the dev mode. It's a can of worms, making the behavior unpredictable at best, and is dangerous in production.
The production-level solutions usually need some external orchestration that would spawn new processes and eventually phase out the old ones gradually.

**Reply by mgvs** [reply]
Yes, agree, 99% of cases. But as I said I have dynamic api based on content of directory. And indeed there is only one way to use external clever load-balancers to shut/start new instance without service downtime.
Anyway thank you for the reply

**Reply by Dreamsorcerer** [reply]
For a use case like that, can't you just use a variable resource?
https://docs.aiohttp.org/en/stable/web_quickstart.html#variable-resources

**Reply by webknjaz** [reply]
While that's the solution, I wouldn't recommend implementing the thing in the first place. The thing described very much sounds like shooting oneself in a foot and opening up a service to security vulnerabilities. It doesn't take into account how Python caches loaded modules, for example, and would lead to untraceable weird in-memory states that only exist in one place in runtime with likely zero ways to reproduce ever again.

**Reply by Dreamsorcerer** [reply]
It wasn't clear to me what the implementation is doing. I was assuming that there is some standard code, but the API routes need to be changed to match a directory structure dynamically. It sounds like you're assuming that modules are being reloaded with new code from disk at runtime. If the latter, that certainly sounds like a crazy thing to do.

**Reply by mgvs** [reply]
I have api server (aiohttp-swagger3) and git repository where users upload their jschema files. So I wrote code which makes /api endpoints with attributes defined at those jschema files. I.e. I get /api/directory_name and fields defined at jschema.
Then I want to use webhook trigger from gitlab to tell api-server re-read repo and refresh dynamic-endpoints.

So it's not about crazy-python code, it's about swagger form for the users

**Reply by webknjaz** [reply]
You probably want a middleware: https://docs.aiohttp.org/en/stable/web_advanced.html#middlewares
